### PR TITLE
Do not crash later on if we are unable to roundtrip a symbol to OOP.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1178861

### CODE REVIEW

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs [hunk 1]
```diff
@@ -75,7 +75,15 @@ public async Task OnDefinitionFoundAsync(SerializableSymbolAndProjectId definiti
...
                 {
-                    symbol = _definitionMap[definition];
+                    // The definition may not be in the map if we failed to map it over using TryRehydrateAsync in OnDefinitionFoundAsync.
+                    // Just ignore this reference.  Note: while this is a degraded experience:
+                    //
+                    // 1. TryRehydrateAsync logs an NFE so we can track down while we're failing to dountrip the
+                    //    definition so we can track down that issue.
+                    // 2. NFE'ing and failing to show a result, is much better than NFE'ing and then crashing
+                    //    immediately afterwards.
+                    if (!_definitionMap.TryGetValue(definition, out symbol))
```
Comment (mavasani) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs:85]: Is it possible to add a unit test?

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs [hunk 2]
```diff
@@ -75,7 +75,15 @@ public async Task OnDefinitionFoundAsync(SerializableSymbolAndProjectId definiti
                 ISymbol symbol;
                 lock (_gate)
                 {
-                    symbol = _definitionMap[definition];
+                    // The definition may not be in the map if we failed to map it over using TryRehydrateAsync in OnDefinitionFoundAsync.
+                    // Just ignore this reference.  Note: while this is a degraded experience:
+                    //
+                    // 1. TryRehydrateAsync logs an NFE so we can track down while we're failing to dountrip the
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs:81]:
```suggestion
                    // 1. TryRehydrateAsync logs an NFE so we can track down while we're failing to roundtrip the
```

**Comment by jinujoseph** [comment]
Can we port this for 16.8.Preview3 ?

**Comment by mavasani** [comment]
@CyrusNajmabadi @sharwell @tmat This fix by itself seems fine. However, there seems to be a much bigger problem with OOP here based on the feedback comments:

> While the error list showed stale results, I found this in the ServiceHub logs: <%stack_trace_with_this_crash%>.
> Closing VS and re-opening gets the error list current, but only for a few minutes before the problem repros.

The above behavior described by @AArnott in the feedback ticket matches what I have started seeing periodically on latest dogfood builds, where executing a FAR command goes into an infinite progress bar and everything supposed to work in OOP, like diagnostics, stops working. I believe @davkean also described similar behavior in Teams.

How can a crash in a specific feature like FAR cause all OOP functionality to stop working? If it does crash OOP, why don't we show a gold bar + treat it similar to user setting the in-proc option so features try to run in-proc? Can we file a separate issue to track this?

**Comment by tmat** [comment]
@mavasani There are known issues with handling errors in OOP. I'm working on addressing that. We don't need another issue to track it.

**Comment by mavasani** [comment]
> @mavasani There are known issues with handling errors in OOP. I'm working on addressing that. We don't need another issue to track it.

Great, thanks! As long as we are aware about these, no concerns.

File: src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Workspaces/Core/Portable/FindSymbols/SymbolFinder.FindReferencesServerCallback.cs:85]: Test added in https://github.com/dotnet/roslyn/pull/47343

**Comment by CyrusNajmabadi** [comment]
> How can a crash in a specific feature like FAR cause all OOP functionality to stop working?

Currently because we have no insulation at the feature level.  IF a feature entrypoint goes down, oop goes down.  We can def attempt to make ourselves resilient here (similar to how we guard against analyzer exceptions).  I will file issue on this.

>  If it does crash OOP, why don't we show a gold bar + treat it similar to user setting the in-proc option so features try to run in-proc? 

That seems somewhat reasonable as well.  Though my preference is to just disable the feature rather than switching modes (which could then just tank performance).

**Comment by CyrusNajmabadi** [comment]
@tmat Do the existing issue track getting feature-isolation in oop so taht if one feature dies, it is isolated from the rest?  If so, great.  If not, i can file that.

**Comment by CyrusNajmabadi** [comment]
@jinujoseph This has been rebased.

**Review by davidwengier — APPROVED** [review]

**Pull request merged**